File: java/src/main/java/Inverse.java
import static java.math.RoundingMode.UNNECESSARY;

import inverse.Price;
import java.math.BigDecimal;

class Inverse {

  private static final BigDecimal VAT = BigDecimal.valueOf(1.2);

  Price addTaxes(Price priceWithoutTaxes) {
    return priceWithoutTaxes.withMappedValue(value -> value.multiply(VAT));
  }

  Price removeTaxes(Price priceWithTaxes) {
    return priceWithTaxes.withMappedValue(value -> value.divide(VAT, UNNECESSARY));
  }

}

File: haskell/src/Inverse.hs
module Inverse where

import           Data.Scientific (Scientific)
import           GHC.Generics    (Generic)

data Price = Price
  { _value :: PriceValue
  , _type  :: PriceType
  } deriving (Show, Eq, Generic)

data PriceType
  = Standard
  | Discount
  deriving (Show, Eq, Generic)

type PriceWithoutTaxes = Price

type PriceWithTaxes = Price

newtype PriceValue = PriceValue
  { _scientificValue :: Scientific
  } deriving (Show, Eq, Generic, Num, Fractional)

tva :: PriceValue
tva = PriceValue 1.2

addTaxes :: PriceWithoutTaxes -> PriceWithTaxes
addTaxes price = price {_value = _value price * tva}

removeTaxes :: PriceWithTaxes -> PriceWithoutTaxes
removeTaxes price = price {_value = _value price / tva}
